File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    stack = []
    
    for n in number:
        while stack and stack[-1] < n and k > 0:
            stack.pop()
            k -= 1
        stack.append(n)
    
    while k > 0:
        stack.pop()
        k -= 1
    
    return ''.join(stack)

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.Stack;
import java.util.stream.Collectors;

class Solution {
    public String solution(String number, int k) {
        Stack<Character> stack = new Stack<>();
        
        for (char n : number.toCharArray()) {
            while (!stack.isEmpty() && stack.peek() < n && k > 0) {
                stack.pop();
                k--;
            }
            stack.add(n);
        }
        
        while (k > 0) {
            stack.pop();
            k--;
        }
        
        return stack.stream()
                .map(String::valueOf)
                .collect(Collectors.joining(""));
    }
}